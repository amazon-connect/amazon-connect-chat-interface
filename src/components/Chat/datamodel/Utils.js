// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import {
  Status,
  Direction,
  ItemDetails,
  TransportDetails,
  ContentType,
  PARTICIPANT_TYPES
} from "./Model";

function isRecognizedEvent(eventName) {
  var values = Object.values(ContentType.EVENT_CONTENT_TYPE);
  for (var index = 0; index < values.length; index++) {
    if (values[index] === eventName) {
      return true;
    }
  }
  return false;
}

function getContent(item) {
  var content = {};
  content.data = item.Content;
  content.type = item.ContentType;
  return content;
}

function createItemFromIncoming(item, thisParticipant) {
  var transcriptItem = new ItemDetails();
  var transportDetails = new TransportDetails();
  transcriptItem.id = item.Id;
  transcriptItem.type = item.Type;
  transcriptItem.content = getContent(item);
  transcriptItem.displayName = item.DisplayName;
  transcriptItem.participantId = item.ParticipantId;
  transcriptItem.participantRole = item.ParticipantRole;
  transportDetails.direction = 
      item.ParticipantRole === PARTICIPANT_TYPES.CUSTOMER
        ? Direction.Outgoing
        : Direction.Incoming;
  transportDetails.sentTime = 
    new Date(item.AbsoluteTime).getTime() / 1000;
  transportDetails.status = Status.SendSuccess;
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.version = 0;
  transcriptItem.Attachments = item.Attachments;
  return transcriptItem;
}

function createTranscriptItemFromSuccessResponse(oldTranscriptItem, response) {
  const newTranscriptItem = new ItemDetails();
  Object.assign(newTranscriptItem, oldTranscriptItem);

  if(response.data && response.data.Id){
    newTranscriptItem.id = response.data.Id;
  }

  newTranscriptItem.transportDetails = {
    ...oldTranscriptItem.transportDetails,
    status: Status.SendSuccess,
    sentTime: new Date(response.data.AbsoluteTime || Date.now()).getTime() / 1000
  };
  return newTranscriptItem;
}

function createOutgoingTranscriptItem(type, content, participant){
  const transcriptItem = new ItemDetails();
  const transportDetails = {};
  transcriptItem.type = type;
  transcriptItem.content = content;
  transcriptItem.participantId = participant.participantId;
  transcriptItem.participantRole = PARTICIPANT_TYPES.CUSTOMER;
  transcriptItem.displayName = participant.displayName;
  transportDetails.status = Status.Sending;
  transportDetails.direction = Direction.Outgoing;
  transportDetails.sentTime = _timestampNow();
  transcriptItem.transportDetails = transportDetails;
  transcriptItem.id = _generateLocalId();
  transcriptItem.version = 0;
  return transcriptItem;
}

function createFailedItem(item, sentTime) {
  const clonedItem = new ItemDetails(item);
  clonedItem.transportDetails.status = Status.SendFailed;
  clonedItem.transportDetails.sentTime = sentTime;
  clonedItem.version = clonedItem.version + 1;
  return clonedItem;
}

function _generateLocalId() {
  var dt = new Date().getTime();
  var uuid = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
    c
  ) {
    var r = (dt + Math.random() * 16) % 16 | 0;
    dt = Math.floor(dt / 16);
    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
  });
  return uuid;
}

function _timestampNow() {
  var date = new Date();
  var timestamp = date.getTime();
  return timestamp / 1000;
}

function createTypingParticipant(typingDataItem, thisParticipantId) {
  var participantTypingDetails = {};
  participantTypingDetails.participantId = typingDataItem.ParticipantId;
  participantTypingDetails.displayName = typingDataItem.DisplayName;
  var direction =
    thisParticipantId === typingDataItem.ParticipantId
      ? Direction.Outgoing
      : Direction.Incoming;
  participantTypingDetails.direction = direction;
  return participantTypingDetails;
}

function isAttachmentContentType(contentType) {
  return contentType && Object.values(ContentType.ATTACHMENT_CONTENT_TYPE).includes(contentType.toLowerCase());
}

var modelUtils = {
  createItemFromIncoming: createItemFromIncoming,
  createOutgoingTranscriptItem: createOutgoingTranscriptItem,
  createFailedItem: createFailedItem,
  createTypingParticipant: createTypingParticipant,
  isRecognizedEvent: isRecognizedEvent,
  createTranscriptItemFromSuccessResponse: createTranscriptItemFromSuccessResponse,
  isAttachmentContentType: isAttachmentContentType,
};

export { modelUtils };