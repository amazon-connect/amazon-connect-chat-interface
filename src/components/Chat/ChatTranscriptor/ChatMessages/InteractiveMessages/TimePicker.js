// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React, { useState } from "react";
import PT from "prop-types";
import { Button } from "connect-core";
import styled from "styled-components";
import { ResponsesSection, HeaderText } from "../InteractiveMessage";
import { truncateStrFromCharLimit } from "../../../../../utils/helper";
import { InteractiveMessageType } from "../../../datamodel/Model";

const NUM_TIMESLOTS_PER_PAGE = 3;

const DatePicker = styled.div`
  background-color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  padding: ${({ theme }) => theme.globals.basePadding};
  
  span {
    padding: 0 5px;
    flex-grow: 2;
  }
`;
const DatePickerButton = styled(Button)`
  border: ${({ theme }) => theme.globals.baseBorder};
  padding: 0;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  
  &[disabled] {
    opacity: 0.3 !important;
  }
  
  &>div {
    display: flex;
  }
`;

DatePicker.PrevDateButton = styled(DatePickerButton)`
  align-self: flex-start;
`;

DatePicker.NextDateButton = styled(DatePickerButton)`
  align-self: flex-end;
`;

const ChevronIcon = styled.div`
  font-size: 0;
  transform: rotate(${props => props.direction === "left" ? "180" : "0"}deg);
  
  svg {
    width: 8px;
    height: 12px;
  }
`;

const TimeslotsList = styled.div`
  padding: 0 ${({ theme }) => theme.spacing.large};
  
  button {
    margin-top: ${({ theme }) => theme.globals.basePadding};
    width: 100%;
    max-width: none;
    border-radius: ${({ theme }) => theme.spacing.mini};
    white-space: pre-line;
    
    &[data-selected='true'] { 
      box-shadow: none;
      border: 1px solid ${({ theme }) => theme.color.highlightColor};
    }
  }
`;

const TimeslotControls = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${({ theme }) => theme.globals.basePadding} ${({theme }) => theme.spacing.large };
  
  button{
      border-radius: ${({ theme }) => theme.spacing.mini};
	  padding: ${({ theme }) => theme.globals.basePadding};
	     
	  &[disabled] {
	    opacity: 0.3 !important;
	  }
  }
`;

const PrevTimeslotsButton = styled.button`
  background-color: #fff;
  align-self: flex-start;
`;

const NextTimeslotsButton = styled.button`
  background-color: #fff;
  align-self: flex-end;
`;

const ConfirmControlsButton = styled.button`
  background-color: #3F5773;
  color: #fff !important;
`;

const ConfirmSelectionButton = styled(ConfirmControlsButton)`
  flex-grow: 2;
`;
const ResetSelectionButton = styled(ConfirmControlsButton)`
  font-size: 0;
  margin-right: ${({theme}) => theme.spacing.mini};
  svg {
    width: ${({theme}) => theme.fontsSize.small};
    height: ${({theme}) => theme.fontsSize.small};
  }
`;

function getLocale(){
  return (navigator.languages && navigator.languages.length > 0 ? navigator.languages[0] : navigator.language) || 'en-US';
}

function TimeslotButton({timeslot, onClick, onKeyPress, selected }) {
    const { date } = timeslot;
    const parsedDate = new Date(date);

    const start = parsedDate.toLocaleTimeString(
      `${getLocale()}`,
      {
        hour: "numeric",
        minute: "numeric",
        timeZoneName: "short"
      }
    );

    return (
        <Button selected={selected} value={date} onKeyPress={onKeyPress} onClick={onClick}>
            {`${start}` }
        </Button>
    );
}

TimePicker.propTypes = {
  content: PT.object.isRequired,
  addMessage: PT.func.isRequired
};

function Chevron({ direction }){
  return (
    <ChevronIcon direction={direction}>
      <svg width="8px" height="12px" viewBox="0 0 8 12" xmlns="http://www.w3.org/2000/svg">
        <title>Path</title>
        <g id="Page-1" stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
          <polygon id="Path" fill="currentColor" fillRule="nonzero" points="2 0 0.59 1.41 5.17 6 0.59 10.59 2 12 8 6"></polygon>
        </g>
      </svg>
    </ChevronIcon>
  )
}

export default function TimePicker({ content, addMessage }) {
  const { title: inputTitle, subtitle: inputSubtitle, timeZoneOffset, timeslots } = content;
  const title = truncateStrFromCharLimit(inputTitle, InteractiveMessageType.TIME_PICKER, "titleCharLimit");
  const subtitle = truncateStrFromCharLimit(inputSubtitle, InteractiveMessageType.TIME_PICKER, "subtitleCharLimit");

  const timeSlotsWithTimeZoneOffset = getTimeslotsWithTimeZoneOffset(timeslots, timeZoneOffset);
  const timeslotsGroupedByDate = getTimeslotsGroupedByDate(timeSlotsWithTimeZoneOffset);
  const availableDates = Object.keys(timeslotsGroupedByDate);

  const [datePageIndex, setDatePageIndex] = useState(0);
  const selectedDate = availableDates[datePageIndex];
  const timeslotsForSelectedDate = timeslotsGroupedByDate[selectedDate];

  const [selectedTimeslot, setSelectedTimeslot] = useState(null);
  const [timeslotPageIndex, setTimeslotPageIndex] = useState(0);
  const timeslotStartIndex = timeslotPageIndex * NUM_TIMESLOTS_PER_PAGE;
  const timeslotNextPageStartIndex = timeslotStartIndex + NUM_TIMESLOTS_PER_PAGE;
  const visibleTimeslotsForSelectedDate = timeslotsForSelectedDate.slice(timeslotStartIndex, Math.min(timeslotsForSelectedDate.length, timeslotNextPageStartIndex));

  function showEarlierDate() {
    changeDate(datePageIndex - 1);
  }

  function showLaterDate() {
    changeDate(datePageIndex + 1);
  }

  function changeDate(pageIndex) {
    setDatePageIndex(pageIndex);
    setTimeslotPageIndex(0);
    setSelectedTimeslot(null);
  }

  function showEarlierTimeslots() {
    setTimeslotPageIndex(timeslotPageIndex - 1);
  }

  function showLaterTimeslots() {
    setTimeslotPageIndex(timeslotPageIndex + 1);
  }

  function onTimeslotSelect(e) {
    setSelectedTimeslot(e.currentTarget.value);
  }

  function resetSelection(){
    setSelectedTimeslot(null);
  }

  function confirmSelection() {
    addMessage({ text: selectedTimeslot });
  }

  function renderTimeslot(timeslot) {
    const selected = new Date(selectedTimeslot).getTime() === new Date(timeslot.date).getTime();
    return <TimeslotButton selected={selected} key={`timeslot${timeslot.date}`} timeslot={timeslot} onKeypress={onTimeslotSelect} onClick={onTimeslotSelect}/>;
  }
    const dateString = new Date(selectedDate).toLocaleDateString(getLocale(),
      {
        weekday: "long",
        month: "long",
        day: "numeric",
      });
  const showDateControls = availableDates.length > 0;
  const showTimeslotPaginationButtons = selectedTimeslot == null && timeslotsForSelectedDate.length > NUM_TIMESLOTS_PER_PAGE;

  return (
    <>
      <HeaderText title={title} subtitle={subtitle} />
      <ResponsesSection>
        <DatePicker>
          {showDateControls &&
          <DatePicker.PrevDateButton disabled={datePageIndex === 0} onKeyPress={showEarlierDate} onClick={showEarlierDate} data-testid={`time-picker-prev-date-button`}>
            <Chevron direction={"left"}/>
          </DatePicker.PrevDateButton>}
          <span>{dateString}</span>
          {showDateControls &&
          <DatePicker.NextDateButton disabled={datePageIndex === availableDates.length - 1} onClick={showLaterDate} data-testid={`time-picker-next-date-button`}>
            <Chevron direction={"right"}/>
          </DatePicker.NextDateButton>}
        </DatePicker>
        <TimeslotsList>
          {visibleTimeslotsForSelectedDate.map(renderTimeslot)}
        </TimeslotsList>
        <TimeslotControls>
          {selectedTimeslot != null &&
          <>
            <ResetSelectionButton onClick={resetSelection} data-testid={`time-picker-reset-selection-button`}>
              <svg viewBox="0 0 13 13" xmlns="http://www.w3.org/2000/svg" fill="currentColor">
                <path d="M13 1.3L11.7 0 6.5 5.2 1.3 0 0 1.3l5.2 5.2L0 11.7 1.3 13l5.2-5.2 5.2 5.2 1.3-1.3-5.2-5.2z" fillRule="evenodd"/>
              </svg>
            </ResetSelectionButton>
            <ConfirmSelectionButton disabled={selectedTimeslot == null} onClick={confirmSelection} data-testid={`time-picker-confirm-selection-button`}>Confirm</ConfirmSelectionButton>
          </>}
            {showTimeslotPaginationButtons &&
            <>
            <PrevTimeslotsButton disabled={timeslotPageIndex === 0}
                                 onClick={showEarlierTimeslots}>Earlier</PrevTimeslotsButton>
            <NextTimeslotsButton disabled={timeslotNextPageStartIndex >= timeslotsForSelectedDate.length}
                  onClick={showLaterTimeslots}>Later</NextTimeslotsButton>
            </>}
        </TimeslotControls>
      </ResponsesSection>
    </>
  );
}

function getTimeslotsGroupedByDate(timeslots){
  //iso dates are lexicographically sortable, may hve to switch to comparing by date objects though
  timeslots.sort((a, b) => a.date.localeCompare(b.date));
  return timeslots.reduce((slotMap, slot) => {
    const msToMidnightOfDate = new Date(slot.date).setHours(0, 0, 0, 0);
    const dateKey = new Date(msToMidnightOfDate).toDateString();
    if(!slotMap[dateKey]){
      slotMap[dateKey] = [];
    }
    slotMap[dateKey].push(slot)
    return slotMap;
  }, {});
}

function applyTimeZoneOffset(date, timeZoneOffset) {
    if (timeZoneOffset) {
        const dateStringToMilli = new Date(date).getTime();
        const timeZoneOffsetMinutesToMilli = 60000 * timeZoneOffset;
        const totalMilliWithTimeZoneOffset = dateStringToMilli + timeZoneOffsetMinutesToMilli;
        return new Date(totalMilliWithTimeZoneOffset).toISOString().slice(0, -2) + "+00:00";
    } else {
        // timeZoneOffset is optional. If none is provided, then return just the date.
        return date;
    }
}

function getTimeslotsWithTimeZoneOffset(timeslots, timeZoneOffset) {
    const timeslotsWithTimeZoneOffset = timeslots.map((ts) => {
        return {
            ...ts,
            date: applyTimeZoneOffset(ts.date, timeZoneOffset),
        }
    });
    return timeslotsWithTimeZoneOffset
}