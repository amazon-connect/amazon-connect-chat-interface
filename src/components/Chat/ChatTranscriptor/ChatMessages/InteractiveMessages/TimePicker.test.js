// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from 'react';
import { screen, render, fireEvent } from "@testing-library/react"
import { InteractiveMessageType } from "../../../datamodel/Model";
import { ThemeProvider } from "../../../../../theme";
import TimePicker from "./TimePicker";
import * as helpers from '../../../../../utils/helper';

const TIME_PICKER_CONSTRAINTS = helpers.INTERACTIVE_MESSAGE_CONSTRAINTS[InteractiveMessageType.TIME_PICKER];

const mockTimePickerContent = {
    title: "TimePickerTitle",
    subtitle: "TimePickerSubtitle",
    timeslots: [
        {
            date: "2020-10-31T17:00+00:00",
            duration: "60"
        },
        {
            date: "2020-11-15T13:00+00:00",
            duration: "60"
        },
        {
            date: "2020-11-15T16:00+00:00",
            duration: "60"
        }
    ]
};

const mockTimePickerContentWithTimezoneOffset = {
    title: "TimePickerTitle https://www.amazon.com/",
    subtitle: "TimePickerSubtitle https://www.amazon.com/",
    timeZoneOffset: "-30",
    timeslots: [
        {
            date: "2020-10-31T17:00+00:00",
            duration: "60"
        },
        {
            date: "2020-11-15T13:00+00:00",
            duration: "60"
        },
        {
            date: "2020-11-15T16:00+00:00",
            duration: "60"
        }
    ]
};

describe("<TimePicker />", () => {
    let mockProps;

    function renderElement(props) {
        return render(
            <ThemeProvider>
                <TimePicker {...props}/>
            </ThemeProvider>
        );
    }

    beforeEach(()=>{
        const addMessage = jest.fn().mockResolvedValue(undefined);
        mockProps = {content: mockTimePickerContent, addMessage: addMessage};
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("Style should match the snapshot", () => {
        const mockTimePicker = renderElement(mockProps);
        expect(mockTimePicker).toMatchSnapshot();

        expect(screen.getByText("TimePickerTitle")).toBeDefined()
        expect(screen.getByText("TimePickerSubtitle")).toBeDefined();
    });

    it("should be able to select a date", () => {
        renderElement(mockProps);
        expect(screen.getByText("Saturday, October 31")).toBeDefined();
        expect(screen.getByText("5:00 PM UTC")).toBeDefined();
        fireEvent.click(screen.getByTestId("time-picker-next-date-button"));
        expect(screen.getByText("Sunday, November 15")).toBeDefined();
        expect(screen.getByText("1:00 PM UTC")).toBeDefined();
        expect(screen.getByText("4:00 PM UTC")).toBeDefined();
    });

    it("should be able to select and reset time", () => {
        renderElement(mockProps);
        fireEvent.click(screen.getByText("5:00 PM UTC"));
        expect(screen.getByTestId("time-picker-reset-selection-button")).toBeDefined();
        expect(screen.getByTestId("time-picker-confirm-selection-button")).toBeDefined();
        fireEvent.click(screen.getByTestId("time-picker-reset-selection-button"));
        expect(screen.queryByTestId("time-picker-confirm-selection-button")).toBeNull();
    });

    it("should be able to confirm the time", () => {
        renderElement(mockProps);
        fireEvent.click(screen.getByText("5:00 PM UTC"));
        fireEvent.click(screen.getByTestId("time-picker-confirm-selection-button"));
        expect(mockProps.addMessage).toHaveBeenCalledTimes(1);
        expect(mockProps.addMessage).toHaveBeenCalledWith({text: "2020-10-31T17:00+00:00"});
    });

    it("should be able to apply specified timezone offset", () => {
        mockProps.content = mockTimePickerContentWithTimezoneOffset;
        renderElement(mockProps);
        expect(screen.getByText("Saturday, October 31")).toBeDefined();
        expect(screen.getByText("4:30 PM UTC")).toBeDefined();
        fireEvent.click(screen.getByTestId("time-picker-next-date-button"));
        expect(screen.getByText("Sunday, November 15")).toBeDefined();
        expect(screen.getByText("12:30 PM UTC")).toBeDefined();
        expect(screen.getByText("3:30 PM UTC")).toBeDefined();
    });

    it("should be able to render time options for a specific timezone", () => {
        const mockToLocaleTimeString = jest.spyOn(Date.prototype, "toLocaleTimeString").mockReturnValue("7:00 AM PDT");
        renderElement(mockProps);
        expect(mockToLocaleTimeString).toHaveBeenCalledTimes(1);
        expect(screen.getByText("Saturday, October 31")).toBeDefined();
        expect(screen.getByText("7:00 AM PDT")).toBeDefined();
    });

    it.each([
        ["title", TIME_PICKER_CONSTRAINTS.titleCharLimit],
        ["subtitle", TIME_PICKER_CONSTRAINTS.subtitleCharLimit]
    ])("should truncate a %s over the character limit", (fieldKey, charLimit) => {
        const longTitle = "LongTitle".repeat(150);
        const truncatedTitle = `${longTitle.substring(0, charLimit)}...`;

        const timePickerLongTitle = {
            ...mockTimePickerContent,
            [fieldKey]: longTitle,
        };
        renderElement({ ...mockProps, content: timePickerLongTitle });
        expect(() => screen.getByText(longTitle)).toThrow(
            "Unable to find an element"
        );
        expect(() => screen.getByText(truncatedTitle)).not.toThrow();
    });
});

describe("TimePicker XSS Mitigation", () => {
    let mockProps;
    const mockTimePickerXSSContent = {
      ...mockTimePickerContent,
      title: 'Title with script <script>alert("XSS attack!");</script>',
      subtitle: '<input type="text" value="XSS attack!" onfocus="alert(\'XSS attack!\');"></input>',
    };
    beforeEach(() => {
      mockProps = {
        content: mockTimePickerXSSContent,
        addMessage: jest.fn(),
        templateType: InteractiveMessageType.TIME_PICKER,
      };
    });
    function renderElement(props) {
      return render(
        <ThemeProvider>
            <TimePicker {...props}/>
        </ThemeProvider>
      );
    }

    it("should use DOMPurify to mitigate malicious XSS input", () => {
      renderElement(mockProps);
      expect(screen.getByText("Title with script")).toBeDefined();
      expect(screen.getByText("<input value=\"XSS attack!\" type=\"text\">", { exact: false })).toBeDefined();
    });
  });