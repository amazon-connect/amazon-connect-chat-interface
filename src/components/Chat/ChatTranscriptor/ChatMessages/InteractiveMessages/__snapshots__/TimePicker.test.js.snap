// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePicker /> Style should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  color: inherit;
  margin: 0;
}

.c1:last-child {
  margin: 0;
}

.c6 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c6:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c6:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c6:active {
  border-color: #fff #fff #fff;
}

.c6:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c6:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c6 > img,
.c6 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c6:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c6:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c6:active {
  border-color: #fff #fff #fff;
}

.c6:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c6:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c2 {
  font-weight: bold !important;
}

.c3 {
  color: #555555;
}

.c0 {
  padding: 14px;
  text-align: left;
}

.c4 {
  position: relative;
  white-space: pre-line;
  border-radius: 6px;
}

.c5 {
  background-color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  padding: 10px;
}

.c5 span {
  padding: 0 5px;
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
}

.c7 {
  border: 1px solid #DDDDDD;
  padding: 0;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c7[disabled] {
  opacity: 0.3 !important;
}

.c7>div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  border: 1px solid #DDDDDD;
  padding: 0;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

.c9[disabled] {
  opacity: 0.3 !important;
}

.c9>div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  font-size: 0;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c8 svg {
  width: 8px;
  height: 12px;
}

.c10 {
  font-size: 0;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c10 svg {
  width: 8px;
  height: 12px;
}

.c11 {
  padding: 0 22px;
}

.c11 button {
  margin-top: 10px;
  width: 100%;
  max-width: none;
  border-radius: 6px;
  white-space: pre-line;
}

.c11 button[data-selected='true'] {
  box-shadow: none;
  border: 1px solid #398ADA;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 22px;
}

.c12 button {
  border-radius: 6px;
  padding: 10px;
}

.c12 button[disabled] {
  opacity: 0.3 !important;
}

<body>
    <div>
      <div
        class="c0"
      >
        <p
          class="c1 c2"
          style="margin: 0px;"
        >
          TimePickerTitle
        </p>
        <p
          class="c1 c3"
          style="margin: 0px;"
        >
          TimePickerSubtitle
        </p>
      </div>
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <button
            class="c6 c7"
            data-selected="false"
            data-testid="time-picker-prev-date-button"
            disabled=""
            type="default"
          >
            <div
              class="c8"
              direction="left"
            >
              <svg
                height="12px"
                viewBox="0 0 8 12"
                width="8px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  Path
                </title>
                <g
                  fill="none"
                  fill-rule="evenodd"
                  id="Page-1"
                  stroke="none"
                  stroke-width="1"
                >
                  <polygon
                    fill="currentColor"
                    fill-rule="nonzero"
                    id="Path"
                    points="2 0 0.59 1.41 5.17 6 0.59 10.59 2 12 8 6"
                  />
                </g>
              </svg>
            </div>
          </button>
          <span>
            Saturday, October 31
          </span>
          <button
            class="c6 c9"
            data-selected="false"
            data-testid="time-picker-next-date-button"
            type="default"
          >
            <div
              class="c10"
              direction="right"
            >
              <svg
                height="12px"
                viewBox="0 0 8 12"
                width="8px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  Path
                </title>
                <g
                  fill="none"
                  fill-rule="evenodd"
                  id="Page-1"
                  stroke="none"
                  stroke-width="1"
                >
                  <polygon
                    fill="currentColor"
                    fill-rule="nonzero"
                    id="Path"
                    points="2 0 0.59 1.41 5.17 6 0.59 10.59 2 12 8 6"
                  />
                </g>
              </svg>
            </div>
          </button>
        </div>
        <div
          class="c11"
        >
          <button
            class="c6"
            data-selected="false"
            type="default"
            value="2020-10-31T17:00+00:00"
          >
            5:00 PM UTC
          </button>
        </div>
        <div
          class="c12"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-fAjcbJ bJnSGf"
    >
      <p
        class="sc-bZQynM fraGhu sc-fMiknA guHZOl"
        style="margin: 0px;"
      >
        TimePickerTitle
      </p>
      <p
        class="sc-bZQynM fraGhu sc-dVhcbM fmjyml"
        style="margin: 0px;"
      >
        TimePickerSubtitle
      </p>
    </div>
    <div
      class="sc-caSCKo eMgFmA"
    >
      <div
        class="sc-cHGsZl kpeViL"
      >
        <button
          class="sc-gzVnrw fgSGc sc-TOsTZ sc-kgAjT cxxbiC"
          data-selected="false"
          data-testid="time-picker-prev-date-button"
          disabled=""
          type="default"
        >
          <div
            class="sc-ksYbfQ kgnwri"
            direction="left"
          >
            <svg
              height="12px"
              viewBox="0 0 8 12"
              width="8px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Path
              </title>
              <g
                fill="none"
                fill-rule="evenodd"
                id="Page-1"
                stroke="none"
                stroke-width="1"
              >
                <polygon
                  fill="currentColor"
                  fill-rule="nonzero"
                  id="Path"
                  points="2 0 0.59 1.41 5.17 6 0.59 10.59 2 12 8 6"
                />
              </g>
            </svg>
          </div>
        </button>
        <span>
          Saturday, October 31
        </span>
        <button
          class="sc-gzVnrw fgSGc sc-TOsTZ sc-cJSrbW krSpzL"
          data-selected="false"
          data-testid="time-picker-next-date-button"
          type="default"
        >
          <div
            class="sc-ksYbfQ leEBYp"
            direction="right"
          >
            <svg
              height="12px"
              viewBox="0 0 8 12"
              width="8px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Path
              </title>
              <g
                fill="none"
                fill-rule="evenodd"
                id="Page-1"
                stroke="none"
                stroke-width="1"
              >
                <polygon
                  fill="currentColor"
                  fill-rule="nonzero"
                  id="Path"
                  points="2 0 0.59 1.41 5.17 6 0.59 10.59 2 12 8 6"
                />
              </g>
            </svg>
          </div>
        </button>
      </div>
      <div
        class="sc-hmzhuo cAkzWi"
      >
        <button
          class="sc-gzVnrw fgSGc"
          data-selected="false"
          type="default"
          value="2020-10-31T17:00+00:00"
        >
          5:00 PM UTC
        </button>
      </div>
      <div
        class="sc-frDJqD cxKniI"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
